packages/plugin-inline/Inline.ts

```diff
@@ -0,0 +1,102 @@
+import { JWPlugin } from '../core/src/JWPlugin';
+import { RangeParams } from '../core/src/CorePlugin';
+import { Format } from './Format';
+import { InlineNode } from './InlineNode';
+import { Constructor } from '../utils/src/utils';
+
+export interface FormatParams extends RangeParams {
+    FormatClass: Constructor<Format>;
+}
+
+export class Inline extends JWPlugin {
+    commands = {
+        toggleFormat: {
+            handler: this.toggleFormat.bind(this),
+        },
+    };
+    commandHooks = {
+        setSelection: this.resetFormatCache.bind(this),
+    };
+    /**
+     * When apply format on a collapsed range, cache the calculation of the
+     * format the following property. This value is reset each time the range
+     * change in a document.
+     */
+    formatCache: Format[] = null;
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * Apply the `format` to the selection.
+     *
+     * If the selection is collapsed, set the format on the selection so we know
+     * in the next insert which format should be used.
+     */
+    toggleFormat(params: FormatParams): void {
+        const range = params.range || this.editor.vDocument.selection.range;
+        const FormatClass = params.FormatClass;
+        if (range.isCollapsed()) {
+            if (!this.formatCache) {
+                this.formatCache = this.getCurrentFormats();
+            }
+            const index = this.formatCache.findIndex(f => f instanceof FormatClass);
+            if (index !== -1) {
+                this.formatCache.splice(index, 1);
+            } else {
+                this.formatCache.push(new FormatClass());
+            }
+        } else {
+            const selectedInlines = range.selectedNodes(InlineNode);
+
+            // If every char in the range has the format `FormatClass`, remove
+            // the format for all of them.
+            const allHaveFormat = selectedInlines.every(inline => {
+                return !!inline.formats.find(f => f instanceof FormatClass);
+            });
+            if (allHaveFormat) {
+                for (const inline of selectedInlines) {
+                    const index = inline.formats.findIndex(f => f instanceof FormatClass);
+                    const matchingFormat = inline.formats[index];
+                    for (const key of Object.keys(matchingFormat.attributes)) {
+                        inline.attributes[key] = matchingFormat.attributes[key];
```

Comment:
This is "extending" an existing b with classes, right ?
So:
```<b class"dmo">text</b>[ more text]```
Becomes:
```<b class"dmo">text[ more text]</b>```
Instead of:
```<b class"dmo">text</b><b>[ more text]</b>```

Is this what that is doing ?

If yes then what about:
```<b class"dmo">text</b>[ more ]<b class="chm">text</b>```

Re-reading the code, I'm not sure this is what it does... I think a comment would be great.